Services/AssessmentQuestion/examples/class.exObjQuestionPoolGUI.php

```diff
@@ -0,0 +1,155 @@
+				$backLink = ''; // Initialise with Back Link to Consumers Back-Landing Page
+				
+				$factory = new ilAsqFactory();
+				
+				$questionInstance = $factory->getQuestionInstance($questionId);
+				$questionAuthoringGUI = $factory->getAuthoringCommandInstance($questionInstance);
+				
+				$questionAuthoringGUI->setBackLink($backLink);
+				
+				$DIC->ctrl()->forwardCommand($questionAuthoringGUI);
+		}
+	}
+	
+	/**
+	 * For question listings the ilAsqQuestionFactory provides a factory method to retrieve
+	 * an array of associative question data arrays. This structure can be simply used as
+	 * data structure for any ilTable2 implementation.
+	 */
+	public function showQuestions()
+	{
+		$parentObjectId = 0; // init with question pool object id
+		
+		$factory = new ilAsqFactory();
+		
+		$questionDataArray = $factory->getQuestionDataArray($parentObjectId);
+		
+		/**
+		 * initialise any ilTable2GUI with this data array
+		 * render initialised ilTable2GUI
+		 */
+		
+		$tableGUI = new exQuestionsTableGUI($this, 'showQuestions', '');
+		$tableGUI->setData($questionDataArray);
+		
+		$tableHTML = $tableGUI->getHTML(); // render table
+	}
+	
+	/**
+	 * When a component provides import functionality for assessment questions, it needs to make use of the
+	 * ILIAS QTI service to get any qti xml parsed to an QTI object graph provided by the QTI service.
+	 * 
+	 * To actually import the question as an assessment question the ilAsqQuestion interface method
+	 * fromQtiItem can be used. To retrieve an empty ilAsqQuestion instance, the question type of the
+	 * QtiItem needs to be determined.
+	 * 
+	 * For the question type determination the ilAsqService class provides a corresponding static method.
+	 */
+	public function importQuestions()
+	{
+		$parentObjectId = 0; // init with question pool object id
+		
+		$factory = new ilAsqFactory();
+
+		/**
+		 * parse any qti import xml using the QTI Service and retrieve
+		 * an array containing ilQTIItem instances
+		 */
+		$qtiItems = array(); /* @var ilQTIItem[] $qtiItems */
+		
+		foreach($qtiItems as $qtiItem)
+		{
+			$questionType = ilAsqService::determineQuestionTypeByQtiItem($qtiItem);
+			$questionInstance = $factory->getEmptyQuestionInstance($questionType);
+			
+			$questionInstance->fromQtiItem($qtiItem);
+			$questionInstance->setParentId($parentObjectId);
+			$questionInstance->save();
+		}
+	}
+	
+	/**
+	 * When a component provides export functionality for assessment questions, it needs the ilAsqQuestion
+	 * interface method toQtiXML to retrieve an qti item xml string. Since the QTI service does not support
+	 * to fetch an QTI xml string based on an QTI object graph, the current implementation of returning
+	 * the xml string itself will be kept within the toQtiXML interface method.
+	 * 
+	 * To export one or more assessment questions the ilAsqFactory provides factory methods
+	 * to get single or multiple ilAsqQuestion instances.
+	 */
```

Comment:
For the import/export: This should be wrapped using the concepts at
[https://docu.ilias.de/goto_docu_pg_32894_42.html](https://docu.ilias.de/goto_docu_pg_32894_42.html)

The consumer should not need to deal with qti strings, instead just define a tail dependency like
```
public function getXmlExportTailDependencies($a_entity, $a_target_release, $a_ids) {

$question_ids = ...;
return array ( array("component" => "Services/Question", "entity" => "quest", "ids" => $question_ids));

}

```

Reply:
done ;-)